### PR TITLE
Run G4 workers on their own dedicated threads

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Each G4 worker now runs on its own thread. Each stream has its own thread and when the stream needs to do work it blocks the TBB thread and starts up its dedicated thread.

This should allow SimWatchers to be modified to call esConsumes.
#### PR validation:

Code compiles and a simple workflow runs.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
@makortel FYI

**Comment by Dr15Jones** [comment]
please test

**Comment by civanch** [comment]
@Dr15Jones , I would expect regression for this change or we change random seed?

**Comment by Dr15Jones** [comment]
@civanch
> I would expect regression for this change or we change random seed?

Given the tests run single threaded, I would have expected exact comparisons. I'm making a change now where I properly propagate the random number seed to the new threads. Let's see if that avoids the differences.

**Comment by civanch** [comment]
@Dr15Jones , there is a simple method fully control random seeds. I would suggest to use it in the first place. For that uncomment line 217 and make line 216 and 217 edm::LogWarning.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
The DQM histogram differences are all from the MessageLogger and the reco differences are a known reproducibility problem caused by an uninitialized variable. 
So the tests look good.

File: SimG4Core/Application/plugins/OscarMTProducer.cc [hunk 1]
```diff
@@ -165,7 +177,13 @@ OscarMTProducer::OscarMTProducer(edm::ParameterSet const& p, const OscarMTMaster
   edm::LogVerbatim("SimG4CoreApplication") << "OscarMTProducer is constructed";
 }
 
-OscarMTProducer::~OscarMTProducer() {}
+OscarMTProducer::~OscarMTProducer() {
+  auto token = edm::ServiceRegistry::instance().presentToken();
```
Comment (civanch) [inline SimG4Core/Application/plugins/OscarMTProducer.cc:181]: @Dr15Jones , is it safe calling edm::ServiceRegistry::instance() in a destructor? May be better keeping the pointer as a class member?

Comment (Dr15Jones) [reply SimG4Core/Application/plugins/OscarMTProducer.cc:181]: The framework guarantees that it is safe to use the Service system in module destructors.

Comment (civanch) [reply SimG4Core/Application/plugins/OscarMTProducer.cc:181]: +1

If the framework guarantee the order of destruction, it is fine.

**Comment by Dr15Jones** [comment]
@civanch if you intended to approve the whole pull request (and not just the one comment) you have to do the + and 1 as a top level comment.

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**